File: src/disser_messenger/mesh/lan_frame_codec.py
from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from typing import Any, BinaryIO


MAX_FRAME_BYTES = 256 * 1024
LENGTH_PREFIX_BYTES = 4


@dataclass(frozen=True, slots=True)
class LanFrameEnvelope:
    sender_peer_id: str
    sender_fingerprint: str
    packet: dict[str, Any]
    sender_display_name: str | None = None
    sender_reply_port: int | None = None
    frame_version: int = 1


def encode_packet_frame(packet: dict[str, Any]) -> bytes:
    return _encode_payload(packet)


def encode_envelope_frame(envelope: LanFrameEnvelope) -> bytes:
    payload: dict[str, Any] = {
        "frame_version": envelope.frame_version,
        "sender_peer_id": envelope.sender_peer_id,
        "sender_fingerprint": envelope.sender_fingerprint,
        "packet": envelope.packet,
    }
    if envelope.sender_display_name is not None:
        payload["sender_display_name"] = envelope.sender_display_name
    if envelope.sender_reply_port is not None:
        payload["sender_reply_port"] = envelope.sender_reply_port
    return _encode_payload(payload)


def decode_frame(stream: BinaryIO) -> LanFrameEnvelope:
    payload = _read_payload(stream)
    decoded = json.loads(payload.decode("utf-8"))
    if isinstance(decoded, dict) and isinstance(decoded.get("packet"), dict):
        envelope = LanFrameEnvelope(
            frame_version=int(decoded.get("frame_version", 1)),
            sender_peer_id=str(decoded.get("sender_peer_id", "")),
            sender_fingerprint=str(decoded.get("sender_fingerprint", "")),
            sender_display_name=_optional_str(decoded.get("sender_display_name")),
            sender_reply_port=_optional_int(decoded.get("sender_reply_port")),
            packet=decoded["packet"],
        )
        if envelope.sender_fingerprint != str(envelope.packet.get("sender_fingerprint", "")):
            raise ValueError("LAN frame sender fingerprint does not match packet sender.")
        return envelope

    if not isinstance(decoded, dict):
        raise ValueError("Malformed Kraken packet frame.")
    sender_fingerprint = str(decoded.get("sender_fingerprint", ""))
    return LanFrameEnvelope(
        sender_peer_id=f"lan-{sender_fingerprint}",
        sender_fingerprint=sender_fingerprint,
        packet=decoded,
    )


def _encode_payload(payload: dict[str, Any]) -> bytes:
    raw = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    if len(raw) > MAX_FRAME_BYTES:
        raise ValueError(f"Kraken packet frame exceeds {MAX_FRAME_BYTES} bytes.")
    return struct.pack(">I", len(raw)) + raw


def _read_payload(stream: BinaryIO) -> bytes:
    length_prefix = _read_exact(stream, LENGTH_PREFIX_BYTES)
    length = struct.unpack(">I", length_prefix)[0]
    if length < 1 or length > MAX_FRAME_BYTES:
        raise ValueError("Invalid Kraken packet frame length.")
    return _read_exact(stream, length)


def _read_exact(stream: BinaryIO, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining > 0:
        chunk = stream.read(remaining)
        if not chunk:
            raise ValueError("Unexpected end of Kraken packet frame.")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _optional_str(value: Any) -> str | None:
    if value is None:
        return None
    text = str(value)
    return text if text else None


def _optional_int(value: Any) -> int | None:
    if value is None:
        return None
    return int(value)

File: app-android/app/src/main/java/com/disser/kraken/mesh/LanFrameCodec.kt
package com.disser.kraken.mesh

import com.disser.kraken.invite.InvitePayloadCodec
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

@Serializable
data class LanFrameEnvelope(
    @SerialName("frame_version")
    val frameVersion: Int = 1,
    @SerialName("sender_peer_id")
    val senderPeerId: String,
    @SerialName("sender_fingerprint")
    val senderFingerprint: String,
    @SerialName("sender_display_name")
    val senderDisplayName: String? = null,
    @SerialName("sender_reply_port")
    val senderReplyPort: Int? = null,
    val packet: KrakenPacket,
)

object LanFrameCodec {
    const val MAX_FRAME_BYTES = 256 * 1024
    private const val ACK_BYTE = 0x06
    private const val LENGTH_PREFIX_BYTES = 4

    fun encode(
        packet: KrakenPacket,
        senderPeer: DiscoveredPeer? = null,
        senderReplyPort: Int? = null,
    ): ByteArray {
        val encoded = if (senderPeer == null) {
            MeshPacketCodec.encode(packet)
        } else {
            InvitePayloadCodec.json.encodeToString(
                LanFrameEnvelope(
                    senderPeerId = senderPeer.peerId,
                    senderFingerprint = senderPeer.fingerprint,
                    senderDisplayName = senderPeer.displayName,
                    senderReplyPort = senderReplyPort,
                    packet = packet,
                ),
            )
        }
        val payload = encoded.encodeToByteArray()
        require(payload.size <= MAX_FRAME_BYTES) { "Kraken packet frame exceeds $MAX_FRAME_BYTES bytes." }
        val output = ByteArrayOutputStream(LENGTH_PREFIX_BYTES + payload.size)
        output.write(ByteBuffer.allocate(LENGTH_PREFIX_BYTES).putInt(payload.size).array())
        output.write(payload)
        return output.toByteArray()
    }

    fun decode(input: InputStream): KrakenPacket {
        val lengthBytes = input.readFully(LENGTH_PREFIX_BYTES)
        val length = ByteBuffer.wrap(lengthBytes).int
        require(length in 1..MAX_FRAME_BYTES) { "Invalid Kraken packet frame length." }
        val payload = input.readFully(length).decodeToString()
        return MeshPacketCodec.decode(payload).getOrElse {
            throw IllegalArgumentException("Malformed Kraken packet frame.", it)
        }
    }

    fun decodeEnvelope(input: InputStream): LanFrameEnvelope {
        val payload = readPayload(input)
        return runCatching {
            InvitePayloadCodec.json.decodeFromString<LanFrameEnvelope>(payload)
                .validated()
        }.getOrElse {
            val packet = MeshPacketCodec.decode(payload).getOrElse { error ->
                throw IllegalArgumentException("Malformed Kraken packet frame.", error)
            }
            LanFrameEnvelope(
                senderPeerId = "lan-${packet.senderFingerprint}",
                senderFingerprint = packet.senderFingerprint,
                packet = packet,
            )
        }
    }

    fun writeAck(output: OutputStream) {
        output.write(ACK_BYTE)
        output.flush()
    }

    fun readAck(input: InputStream) {
        val ack = input.read()
        require(ack == ACK_BYTE) { "Kraken packet frame ACK was not received." }
    }

    private fun LanFrameEnvelope.validated(): LanFrameEnvelope {
        require(senderFingerprint == packet.senderFingerprint) {
            "LAN frame sender fingerprint does not match packet sender."
        }
        return this
    }

    private fun readPayload(input: InputStream): String {
        val lengthBytes = input.readFully(LENGTH_PREFIX_BYTES)
        val length = ByteBuffer.wrap(lengthBytes).int
        require(length in 1..MAX_FRAME_BYTES) { "Invalid Kraken packet frame length." }
        return input.readFully(length).decodeToString()
    }

    private fun InputStream.readFully(length: Int): ByteArray {
        val buffer = ByteArray(length)
        var offset = 0
        while (offset < length) {
            val read = read(buffer, offset, length - offset)
            if (read == -1) {
                throw IllegalArgumentException("Unexpected end of Kraken packet frame.")
            }
            offset += read
        }
        return buffer
    }
}
